// msa/lib/menu/views/OrderingMenu.js
// Generated by CoffeeScript 1.9.2
var MenuBuilder, OrderingMenu, _, dom;

MenuBuilder = require("../menubuilder");

dom = require("dom-helper");

_ = require('underscore');

module.exports = OrderingMenu = MenuBuilder.extend({
  initialize: function(data) {
    this.g = data.g;
    this.order = "ID";
    return this.el.style.display = "inline-block";
  },
  setOrder: function(order) {
    this.order = order;
    return this.render();
  },
  render: function() {
    var comps, el, i, len, m;
    this.setName("Ordering");
    this.removeAllNodes();
    comps = this.getComparators();
    for (i = 0, len = comps.length; i < len; i++) {
      m = comps[i];
      this._addNode(m);
    }
    el = this.buildDOM();
    dom.removeAllChilds(this.el);
    this.el.appendChild(el);
    return this;
  },
  _addNode: function(m) {
    var style, text;
    text = m.text;
    style = {};
    if (text === this.order) {
      style.backgroundColor = "#77ED80";
    }
    return this.addNode(text, (function(_this) {
      return function() {
        if (m.precode != null) {
          m.precode();
        }
        _this.model.comparator = m.comparator;
        _this.model.sort();
        return _this.setOrder(m.text);
      };
    })(this), {
      style: style
    });
  },
  getComparators: function() {
    var models, setGaps, setIdent;
    models = [];
    models.push({
      text: "ID ↑",
      comparator: "id"
    });
    models.push({
      text: "ID ↓",
      comparator: function(a, b) {
        return -("" + a.get("id")).localeCompare("" + b.get("id"), [], {
          numeric: true
        });
      }
    });
    models.push({
      text: "Label ↑",
      comparator: "name"
    });
    models.push({
      text: "Label ↓",
      comparator: function(a, b) {
        return -a.get("name").localeCompare(b.get("name"));
      }
    });
    models.push({
      text: "Seq ↑",
      comparator: "seq"
    });
    models.push({
      text: "Seq ↓",
      comparator: function(a, b) {
        return -a.get("seq").localeCompare(b.get("seq"));
      }
    });
    setIdent = (function(_this) {
      return function() {
        return _this.ident = _this.g.stats.identity();
      };
    })(this);
    setGaps = (function(_this) {
      return function() {
        _this.gaps = {};
        return _this.model.each(function(el) {
          var seq;
          seq = el.attributes.seq;
          return _this.gaps[el.id] = (_.reduce(seq, (function(memo, c) {
            if (c === '-') {
              memo++;
            }
            return memo;
          }), 0)) / seq.length;
        });
      };
    })(this);
    models.push({
      text: "Identity ↑",
      comparator: (function(_this) {
        return function(a, b) {
          var val;
          val = _this.ident[a.id] - _this.ident[b.id];
          if (val > 0) {
            return 1;
          }
          if (val < 0) {
            return -1;
          }
          return 0;
        };
      })(this),
      precode: setIdent
    });
    models.push({
      text: "Identity ↓",
      comparator: (function(_this) {
        return function(a, b) {
          var val;
          val = _this.ident[a.id] - _this.ident[b.id];
          if (val > 0) {
            return -1;
          }
          if (val < 0) {
            return 1;
          }
          return 0;
        };
      })(this),
      precode: setIdent
    });
    models.push({
      text: "Gaps ↑",
      comparator: (function(_this) {
        return function(a, b) {
          var val;
          val = _this.gaps[a.id] - _this.gaps[b.id];
          if (val > 0) {
            return 1;
          }
          if (val < 0) {
            return -1;
          }
          return 0;
        };
      })(this),
      precode: setGaps
    });
    models.push({
      text: "Gaps ↓",
      comparator: (function(_this) {
        return function(a, b) {
          var val;
          val = _this.gaps[a.id] - _this.gaps[b.id];
          if (val < 0) {
            return 1;
          }
          if (val > 0) {
            return -1;
          }
          return 0;
        };
      })(this),
      precode: setGaps
    });
    models.push({
      text: "Consensus to top",
      comparator: function(seq) {
        return !seq.get("ref");
      }
    });
    return models;
  }
});
